### ISSUE TITLE
Hateoas version upgrade because of resource leak

### ISSUE THREAD

**Issue by akosratku** [spring-projects/spring-boot#8116]
Spring Boot Version: 1.4.3.RELEASE
Sample Application: https://github.com/akosratku/hateoas-leak

Hello,

I've noticed that in the currently used `hateoas-starter` there is a possible resource leak if someone is building links with the `ControllerLinkBuilder` as seen in the sample application.

The cause of this is that `AnnotatedParametersParameterAccessor` will always create a new `BoundMethodParameter` and therefor a new `TypeDescriptor` which is used in `GenericConversionService` as a part of the `ConverterCacheKey` and in the end the annotations won't match on the compared `TypeDescriptors` since they are compared for reference. 

To make things more complicated the converterCache in `GenericConversionService` is storing soft references for converters so this is only a problem for `NO_OP` converters since they have a hard reference in `GenericConversionService` and `SoftReference`'s documentation states that

```
...As long as the referent of a soft
  reference is strongly reachable, that is, is actually in use, the soft
  reference will not be cleared...
```

This change: https://github.com/spring-projects/spring-framework/commit/e38c020535dc5d6b326d5e22f8a59dc35ba7361a could be the soulution but in hateoas, it is already fixed in 0.23 with this change for `AnnotatedParametersParameterAccessor `: https://github.com/spring-projects/spring-hateoas/commit/4838e9f40c8a1082a3af1fb2d4013493b318a550#diff-414dd03cf53af85583e3e538024bd957 .

Let me know if you need any more information.

Cheers,
Ákos

**Comment by snicoll** [comment]
I am a bit confused. It looks like you are reporting a problem for which there is already a solution. Spring Boot 1.5 provides hateoas 0.23 already. What do you expect us to do here?

**Comment by akosratku** [comment]
I have only seen an RC version of 1.5 and also I think it should be fixed in 1.4 (maybe in 1.4.4?) since it can cause OOM pretty fast (as it did so for a few services of ours).

**Comment by snicoll** [comment]
Both version have the framework fix and `1.5.0` uses hateoas 0.23. We won't upgrade it in a maintenance release but you can override the version if you want.
